### PR TITLE
[ML] Add a file structure determination endpoint

### PR DESCRIPTION
**Pull request by droberts195**
This endpoint accepts an arbitrary file in the request body and
attempts to determine the structure.  If successful it also
proposes mappings that could be used when indexing the file's
contents, and calculates simple statistics for each of the fields
that are useful in the data preparation step prior to configuring
machine learning jobs.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            this.sample = sample;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+            if (linesToSample != null && linesToSample <= 0) {
+                validationException = addValidationError("lines_to_sample must be positive if specified", validationException);
+            }
+            if (sample == null) {
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java]: may want to verify that `sample.length() > 0` as well.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestFileStructureAction.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        } else {
+            throw new ElasticsearchParseException("request body is required");
+        }
+
+        return channel -> client.execute(FileStructureAction.INSTANCE, request, new RestToXContentListener<>(channel));
+    }
+
+    @Override
+    protected Set<String> responseParams() {
+        return Collections.singleton(FileStructure.EXPLAIN);
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestFileStructureAction.java]: This is just for my own edification. 

Does this "magically" take the query parameter the user provides and the passes it to the xcontent builder? I don't see that parameter directly referenced from the incoming request and it has be curious :)

Comment (droberts195) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestFileStructureAction.java]: All the `RestRequest` parameters are automatically passed to the `toXContent()` method - see 
https://github.com/elastic/elasticsearch/blob/6a3adbd93565299c9e2c09b3781a2d051d22550d/server/src/main/java/org/elasticsearch/rest/action/RestToXContentListener.java#L47

This method just stops the parameter that is _only_ used for X-content formatting and not added to the action request from triggering the validation check for unused parameters.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/rest/RestFileStructureAction.java]: TIL
Thanks for the run down!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xpack.core.ml.filestructurefinder.FileStructure;
+
+import java.io.IOException;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class FileStructureAction extends Action<FileStructureAction.Response> {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java]: Just a note that this is different in `6.x` so make sure to fix compilation problems before backporting.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    public static class Response extends ActionResponse implements StatusToXContentObject, Writeable {
+
+        private FileStructure fileStructure;
+
+        public Response(FileStructure fileStructure) {
+            this.fileStructure = fileStructure;
+        }
+
+        public Response() {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java]: If we need this at all, it should not be `public`

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        public void setSample(BytesReference sample) {
+            this.sample = sample;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+            if (linesToSample != null && linesToSample <= 0) {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java]: Should we also check against an upper bound?

**Review by dimitris-athanasiou — COMMENTED** [review]
Looks great! A couple minor comments.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java (hunk 5 above)
Comment (droberts195) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/FileStructureAction.java]: If the uploaded data contains fewer lines then we just sample all of what's been uploaded.

Therefore the number of lines that can be sampled is limited by the size of the ES HTTP receive buffer.

One problem with sampling many lines is the time taken to do the analysis.  At some point I should probably add the option to limit the maximum time allowed.  Some other functionality has this facility, like the Grok ingest processor.  But I'd rather do this in a followup PR.

**Comment by droberts195** [comment]
@dimitris-athanasiou I made the constructor package private rather than public as you suggested.

Also, as we discussed verbally, I added "find" to the beginning of the action name as all REST actions should begin with verbs.

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Review by benwtrent — APPROVED** [review]
👍

**Pull request merged**